tests/finance/paymentsplitter_test.go

```diff
@@ -0,0 +1,245 @@
+					if got := new(big.Int).Sub(after, before); got.Cmp(want) != 0 {
+						t.Errorf("After %T.Release() payee %d balance diff of %d (%d to %d); want diff of %d", split, i, got, before, after, want)
+					}
+				}
+			})
+
+		})
+	}
+}
+
+func TestDeterministicDeploymentAddress(t *testing.T) {
+	sim, dep, events := deploy(t, 1, 0)
+
+	payees := []common.Address{sim.Addr(0)}
+	shares := []*big.Int{big.NewInt(1)}
+
+	for i := byte(0); i < 10; i++ {
+		var salt [32]byte
+		salt[0] = i
+
+		predicted, err := dep.PredictDeploymentAddress(nil, salt)
+		if err != nil {
+			t.Errorf("%T.PredictDeploymentAddress(%#x) error %v", dep, salt, err)
+			continue
+		}
+
+		sim.Must(t, "%T.DeployDeterministic(%#x)", dep, salt)(dep.DeployDeterministic(sim.Acc(0), salt, payees, shares))
+		ev, ok := <-events
+		if !ok {
+			t.Fatal("Event channel closed unexpectedly")
+		}
+
+		if got := ev.ClonedPaymentSplitter; !bytes.Equal(got.Bytes(), predicted.Bytes()) {
+			t.Errorf("%T.DeployDeterministic(%#x) got deployed address %v; want %v as returned by PredictDeploymentAddress()", dep, salt, got, predicted)
+		}
+	}
+}
+
+func ExampleGasSaving() {
+	ctx := context.Background()
+
+	sim, err := ethtest.NewSimulatedBackend(3)
+	if err != nil {
+		fmt.Printf("NewSimulatedBacked(1): %v", err)
+		return
+	}
+
+	_, _, dep, err := DeployPaymentSplitterDeployer(sim.Acc(0), sim)
+	if err != nil {
+		fmt.Printf("DeployPaymentSplitterDeployer(): %v", err)
+		return
+	}
+
+	payees := []common.Address{sim.Addr(0), sim.Addr(1), sim.Addr(2)}
+	one := big.NewInt(1)
+	shares := []*big.Int{one, one, one}
+
+	gas := func(desc string, tx *types.Transaction) uint64 {
+		rcpt, err := bind.WaitMined(ctx, sim, tx)
+		if err != nil {
+			fmt.Printf("bind.WaitMined(%s): %v", desc, err)
+			return 0
+		}
+		return rcpt.GasUsed
+	}
+
+	tx, err := dep.Deploy(sim.Acc(0), payees, shares)
+	if err != nil {
+		fmt.Printf("%T.Deploy(): %v", dep, err)
+		return
+	}
+	proxy := gas("deploy proxy", tx)
+
+	_, tx, _, err = DeployPaymentSplitter(sim.Acc(0), sim, payees, shares)
+	if err != nil {
+		fmt.Printf("DeployPaymentSplitter() error %v", err)
+		return
+	}
+	full := gas("deploy full PaymentSplitter", tx)
+
+	saving := full - proxy
+
+	comma := func(i uint64) string { return humanize.Comma(int64(i)) }
+
+	fmt.Println("Deploy proxy:", comma(proxy), "gas")
+	fmt.Println("Deploy full PaymentSplitter:", comma(full), "gas")
+	fmt.Println("*** Savings ***")
+	fmt.Println("Gas: ", comma(saving))
+
+	const (
+		priceInGwei = 100
+		ethToUSD    = 3500
+	)
+	ethSave := big.NewRat(int64(saving), 1e9/priceInGwei)
+	fmt.Printf("At %d gwei: %s%s\n", priceInGwei, ethSave.FloatString(3), eth.Symbol)
+	fmt.Printf("At $%s: $%s", comma(ethToUSD), ethSave.Mul(ethSave, big.NewRat(ethToUSD, 1)).FloatString(2))
+
+	// Output:
+	// Deploy proxy: 285,128 gas
+	// Deploy full PaymentSplitter: 1,589,010 gas
+	// *** Savings ***
+	// Gas:  1,303,882
+	// At 100 gwei: 0.130Ξ
+	// At $3,500: $456.36
```

Comment:
Wow this is cool! I did not know that go-testing offers automatic output compares.

Reply:
Yup, they're great! Any function that starts with `Example` in tests gets run as a test _and_, even better, gets shown as an example by `godoc`. It won't in this case because the `finance` package isn't a real one, but generally they're very valuable for showing intended usage of code.

https://go.dev/blog/examples